pkg/cloudprovider/providers/gce/cloud/mock/mock.go

```diff
@@ -226,3 +227,168 @@ func InsertAlphaAddressHook(ctx context.Context, key *meta.Key, obj *alpha.Addre
 }
+
+// Map from InstanceGroup key to list of Instances
+type instanceGroupAttributes struct {
+	instanceMap map[meta.Key][]*ga.InstanceWithNamedPorts
+}
+
+// AddInstancesHook mocks adding instances from an InstanceGroup
+func AddInstancesHook(ctx context.Context, key *meta.Key, req *ga.InstanceGroupsAddInstancesRequest, m *cloud.MockInstanceGroups) error {
+	_, err := m.Get(ctx, key)
+	if err != nil {
+		return &googleapi.Error{
+			Code:    http.StatusNotFound,
+			Message: fmt.Sprintf("Key: %s was not found in InstanceGroups", key.String()),
+		}
+	}
+
+	if m.X == nil {
+		m.X = instanceGroupAttributes{
+			instanceMap: make(map[meta.Key][]*ga.InstanceWithNamedPorts),
+		}
+	}
+
+	var attrs instanceGroupAttributes
+	attrs = m.X.(instanceGroupAttributes)
+
+	instances, ok := attrs.instanceMap[*key]
+	if !ok {
+		instances = []*ga.InstanceWithNamedPorts{}
+	}
+
+	for _, instance := range req.Instances {
+		iWithPort := &ga.InstanceWithNamedPorts{
+			Instance: instance.Instance,
+		}
+
+		instances = append(instances, iWithPort)
+	}
+
+	attrs.instanceMap[*key] = instances
+	m.X = attrs
+	return nil
+}
+
+// ListInstancesHook mocks listing instances from an InstanceGroup
+func ListInstancesHook(ctx context.Context, key *meta.Key, req *ga.InstanceGroupsListInstancesRequest, filter *filter.F, m *cloud.MockInstanceGroups) ([]*ga.InstanceWithNamedPorts, error) {
+	_, err := m.Get(ctx, key)
+	if err != nil {
+		return nil, &googleapi.Error{
+			Code:    http.StatusNotFound,
+			Message: fmt.Sprintf("Key: %s was not found in InstanceGroups", key.String()),
+		}
+	}
+
+	if m.X == nil {
+		m.X = instanceGroupAttributes{instanceMap: make(map[meta.Key][]*ga.InstanceWithNamedPorts)}
+	}
+
+	var attrs instanceGroupAttributes
+	attrs = m.X.(instanceGroupAttributes)
+
+	instances, ok := attrs.instanceMap[*key]
+	if !ok {
+		instances = []*ga.InstanceWithNamedPorts{}
+	}
+
+	return instances, nil
+}
+
+// RemoveInstancesHook mocks removing instances from an InstanceGroup
+func RemoveInstancesHook(ctx context.Context, key *meta.Key, req *ga.InstanceGroupsRemoveInstancesRequest, m *cloud.MockInstanceGroups) error {
+	_, err := m.Get(ctx, key)
+	if err != nil {
+		return &googleapi.Error{
+			Code:    http.StatusNotFound,
+			Message: fmt.Sprintf("Key: %s was not found in InstanceGroups", key.String()),
+		}
+	}
+
+	if m.X == nil {
+		m.X = instanceGroupAttributes{
+			instanceMap: make(map[meta.Key][]*ga.InstanceWithNamedPorts),
+		}
+	}
+
+	var attrs instanceGroupAttributes
+	attrs = m.X.(instanceGroupAttributes)
+
+	instances, ok := attrs.instanceMap[*key]
+	if !ok {
+		instances = []*ga.InstanceWithNamedPorts{}
+	}
+
+	for _, instanceToRemove := range req.Instances {
+		for i, instance := range instances {
+			if instanceToRemove.Instance == instance.Instance {
+				// Delete instance from instances without preserving order
+				instances[i] = instances[len(instances)-1]
+				instances = instances[:len(instances)-1]
+				break
```

Comment:
What kind of error does GCP return when the instance isn't a member of the instance group? BadRequest maybe? We should emulate whatever it does.

Reply:
Checked on GCP - it returns a 400 error with the message:  "Resource '[instance url] is not a member of '[instance group url'."